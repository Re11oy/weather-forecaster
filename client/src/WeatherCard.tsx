import React from 'react'
import styled from 'styled-components'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { format as formatDate } from 'date-fns'

import {
  faTemperatureHigh,
  faTemperatureLow,
} from '@fortawesome/free-solid-svg-icons'
import classNames from 'classnames/bind'
import { LocationForecast } from './api'

const MAX_LIMIT_COLOR = '#c06c84'
const MIN_LIMIT_COLOR = '#6a8caf'

const Wrapper = styled.section`
  margin: 1rem auto;
  border-radius: 3px;
  background-color: #fff;
  box-shadow: 1px 1px 4px rgba(0, 0, 0, 0.2);
  padding: 1rem 2rem;

  .title {
    display: flex;
    justify-content: space-between;
    align-items: center;

    span {
      font-weight: 300;
      font-size: 2.25em;
    }

    svg {
      margin-left: 0.5rem;
    }
  }

  .forecasts {
    display: flex;
    justify-content: space-between;
    padding: 1rem 0;

    > div {
      display: flex;
      flex-direction: column;
      align-items: center;
      margin: 0 1rem;
    }

    .temp {
      color: #cbcbcb;
      margin-top: 0.5rem;
    }

    .maxLimit {
      color: ${MAX_LIMIT_COLOR};
    }

    .minLimit {
      color: ${MIN_LIMIT_COLOR};
    }
  }
`

const dateFormat = (date: Date): string => formatDate(date, 'dd MMM')
const dayFormat = (date: Date): string => formatDate(date, 'E')

export const WeatherCard: React.FC<LocationForecast> = ({
  city,
  forecasts,
}) => {
  const isMinimumReached = forecasts.find((f) => f.isMinimumReached)
  const isMaximumReached = forecasts.find((f) => f.isMaximumReached)

  return (
    <Wrapper>
      <div className="title">
        <span>{city}</span>
        <div>
          {isMinimumReached && (
            <FontAwesomeIcon
              icon={faTemperatureLow}
              size="2x"
              color={MIN_LIMIT_COLOR}
              data-testid="min-limit"
            />
          )}
          {isMaximumReached && (
            <FontAwesomeIcon
              icon={faTemperatureHigh}
              size="2x"
              color={MAX_LIMIT_COLOR}
              data-testid="max-limit"
            />
          )}
        </div>
      </div>
      <div className="forecasts">
        {forecasts.map((forecast) => {
          const forecastDate = new Date(forecast.day)
          return (
            <div
              key={`${forecastDate.getTime()}`}
              className={classNames({
                minLimit: forecast.isMinimumReached,
                maxLimit: forecast.isMaximumReached,
              })}
            >
              <span>{dayFormat(forecastDate)}</span>
              <span className="date">{dateFormat(forecastDate)}</span>
              <span
                className={classNames('temp', {
                  minLimit: forecast.isMinimumReached,
                  maxLimit: forecast.isMaximumReached,
                })}
              >
                {Math.round(forecast.temp)}Â°
              </span>
            </div>
          )
        })}
      </div>
    </Wrapper>
  )
}
