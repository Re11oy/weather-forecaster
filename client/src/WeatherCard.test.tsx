import React from 'react'
import { render } from '@testing-library/react'
import { WeatherCard } from './WeatherCard'

describe('WeatherCard', () => {
  const testDate = new Date(2020, 10)

  const getForecast = (isMinimumReached = false, isMaximumReached = false) => ({
    city: 'Helsinki',
    forecasts: [
      {
        day: testDate,
        temp: 10,
        isMaximumReached: isMaximumReached,
        isMinimumReached: isMinimumReached,
      },
    ],
  })

  it('should render card with forecast', () => {
    const { container } = render(<WeatherCard {...getForecast()} />)

    expect(container).toMatchSnapshot()
  })

  it('should should display min temperature alert', () => {
    const rendered = render(<WeatherCard {...getForecast(true, false)} />)

    expect(rendered.getByTestId('min-limit')).toBeInTheDocument()
    expect(rendered.getByText('10°')).toHaveClass('minLimit')
  })

  it('should should display max temperature alert', () => {
    const rendered = render(<WeatherCard {...getForecast(false, true)} />)

    expect(rendered.getByTestId('max-limit')).toBeInTheDocument()
    expect(rendered.getByText('10°')).toHaveClass('maxLimit')
  })
})
