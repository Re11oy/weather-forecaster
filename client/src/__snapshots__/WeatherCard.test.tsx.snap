// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherCard should render card with forecast 1`] = `
<div>
  <section
    class="sc-bdnylx eJRQQx"
  >
    <div
      class="title"
    >
      <span>
        Helsinki
      </span>
      <div />
    </div>
    <div
      class="forecasts"
    >
      <div
        class=""
      >
        <span>
          Sun
        </span>
        <span
          class="date"
        >
          01 Nov
        </span>
        <span
          class="temp"
        >
          10
          Â°
        </span>
      </div>
    </div>
  </section>
</div>
`;
